controllers/databaseengine_controller.go

```diff
@@ -177,11 +189,60 @@ func (r *DatabaseEngineReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, err
 	}
 
+	if requeue, err := r.restartIfNeeded(ctx); err != nil {
+		return ctrl.Result{}, err
+	} else if requeue {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
 	return ctrl.Result{
 		RequeueAfter: timeUntilUnlock,
 	}, nil
 }
 
+// restartIfNeeded checks if the operator pod needs to be restarted.
+// It does so by checking if there are any running DBEngines whose CRDs are not registered with the operator.
+// Returns: [requeue(bool), error].
+func (r *DatabaseEngineReconciler) restartIfNeeded(ctx context.Context) (bool, error) {
+	if r.podSelf.Name == "" || r.podSelf.Namespace == "" {
+		return false, nil
+	}
+	dbEngines := &everestv1alpha1.DatabaseEngineList{}
+	if err := r.List(ctx, dbEngines); err != nil {
+		return false, fmt.Errorf("failed to list DatabaseEngines: %w", err)
+	}
+	unregistered := 0
+	for _, dbEngine := range dbEngines.Items {
+		// Wait until all DB engines are either installed/not installed.
+		// This way we can avoid redundant restarts.
+		if dbEngine.Status.State == "" ||
+			dbEngine.Status.State == everestv1alpha1.DBEngineStateInstalling ||
+			dbEngine.Status.State == everestv1alpha1.DBEngineStateUpgrading {
+			return true, nil
+		}
+		if dbEngine.Status.State == everestv1alpha1.DBEngineStateNotInstalled {
+			continue
+		}
+		group, found := operatorEngineTypeToCRDGroup[dbEngine.Spec.Type]
+		if !found {
+			return false, fmt.Errorf("unknown engine type '%s'", dbEngine.Spec.Type)
+		}
+		// Ideally we would also like to check if all registered controllers are also watching the CRs,
+		// but since that's tricky to accomplish, we will only check if the CRDs are registered to the scheme,
+		// since we typically perform that step along with configuring the watches.
+		if !r.Scheme.IsGroupRegistered(group) {
+			unregistered++
```

Comment:
Why don't we just return here instead of counting?

Reply:
Fixed